File: functions/branch_groups/passwordUtils.js
const AWS = require('aws-sdk');
const _ = require('lodash');

const ssm = new AWS.SSM({apiVersion: '2014-11-06'});
let velocifyPasswords;

function generateAccountPasswords(env, refresh = false) {
    var awsParams = {
        Path: `/${env}/velocifyPasswords`,
        WithDecryption: true,
    };
    return new Promise((resolve, reject) => {
        if (velocifyPasswords && !refresh) {
            console.log("::::: Velocify Passwords Already Generated");
            console.log(velocifyPasswords);
            resolve(velocifyPasswords);
        } else {
            console.log("::::: Generating Velocify Passwords");
            getVelocifyPasswords(awsParams).then(params => {
                let passwords = [];
                _.forIn(params, param => {
                    const key = param.Name.replace(`/${env}/velocifyPasswords/`, "");
                    passwords[key] = param.Value;
                });
                console.log("::::: Velocify Passwords Successfully Generated");
                velocifyPasswords = passwords;
                resolve(velocifyPasswords);
            }).catch(error => {
                console.log("ERROR", error);
                reject(error);
            });
        }
    });
}

function getVelocifyPasswords(params) {
    return new Promise((resolve, reject) => {
        console.log("::::: Downloading and Decrypting Account Passwords");
        ssm.getParametersByPath(params, function (err, data) {
            if (err) {
                console.log("::::: ERROR!!! Unable to download and decrypt Accounts Passwords");
                console.log(err, err.stack);
                reject(err);
            } else {
                console.log("::::: Successfully Downloaded and Decrypted Accounts Passwords");
                const passwords = data.Parameters;
                if (data.NextToken) {
                    params.NextToken = data.NextToken;
                    getVelocifyPasswords(params).then(nextPasswords => {
                        resolve([...nextPasswords, ...passwords]);
                    });
                } else {
                    resolve(passwords);
                }
            }
        });
    });
}

module.exports = generateAccountPasswords;

File: functions/leads_expand_recommendation_string/passwordUtils.py
import boto3

client = boto3.client('ssm')
next_token = None
velocify_passwords = None


def generate_velocify_passwords(env, refresh=False):
    global velocify_passwords

    if velocify_passwords and not refresh:
        print("::::: Velocify Passwords Already Generated")
        print velocify_passwords
        return velocify_passwords

    passwords_dict = {}
    parameter_list = []

    try:
        parameter_list = get_passwords(env)
    except Exception as e:
        print(":::: ERROR: Couldn't Fetch Velocify Passwords")
        print(e)

    print("::::: Generating Velocify Passwords")

    for p in parameter_list:
        key = p['Name'].replace("/{}/velocifyPasswords/".format(env), '')
        passwords_dict[key] = p['Value']

    velocify_passwords = passwords_dict

    print("::::: Succesfully Generated Velocify Passwords")

    return velocify_passwords


def get_passwords(env):
    global next_token
    print("::::: Fetching Velocify Passwords")

    if next_token == None:
        parameters = client.get_parameters_by_path(
            Path="/{}/velocifyPasswords".format(env),
            WithDecryption=True,
            MaxResults=1
        )
    else:
        print "::::: Fetching next set"
        parameters = client.get_parameters_by_path(
            Path="/{}/velocifyPasswords".format(env),
            WithDecryption=True,
            NextToken=next_token,
            MaxResults=1
        )

    if 'NextToken' in parameters:
        next_token = parameters['NextToken']
        next_passwords = get_passwords()
        return parameters['Parameters'] + next_passwords

    return parameters['Parameters']
